std/file.d

```diff
@@ -670,20 +650,210 @@ version(Windows) private void writeImpl(const(char)[] name, const(FSChar)* namez
+
+    Params:
+        prefix = Prefix to the random portion of the file name.
+        suffix = Suffix to the random portion of the file name (e.g. for a file
+                 extension).
+        buffer = Data to populate the file with.
+
+    Returns:
+        The name of the generated file.
+
+    See_Also:
+        $(REF tmpfile,std,stdio)
+  +/
+string createTempFile(const void[] buffer = null) @safe
+{
+    return createTempFile(null, null, buffer);
+}
+
+/// Ditto
+string createTempFile(const(char)[] prefix, const(char)[] suffix, const void[] buffer = null) @trusted
+{
+    import std.path : absolutePath, baseName, buildPath, dirName;
+
+    static string genTempName(const(char)[] prefix, const(char)[] suffix)
+    {
+        import std.ascii : digits, letters;
+        import std.random : rndGen;
+        import std.range : chain;
+        import std.string : representation;
+
+        auto name = new char[](prefix.length + 15 + suffix.length);
+        name[0 .. prefix.length] = prefix;
+        name[$ - suffix.length .. $] = suffix;
+
+        rndGen.popFront();
+        auto chars = chain(letters.representation, digits.representation);
+        foreach (ref c; name[prefix.length .. $ - suffix.length])
+        {
+            c = chars[rndGen.front % chars.length];
+            rndGen.popFront();
+        }
+
+        return buildPath(tempDir, name);
+    }
+
+    // Once should be enough given the randomness, but we'll give it three
+    // times to be safe.
+    foreach (i; 0 .. 3)
+    {
+        auto filename = genTempName(prefix, suffix);
+
+        version(Posix)
+        {
+            // O_EXCL does not support symbolic links.
+            if (filename.dirName.isSymlink)
+                filename = buildPath(filename.dirName.readLink(), filename.baseName);
+
+            auto fd = open(tempCString!FSChar(filename), O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
+            if (fd == -1)
+            {
+                // Temporary debugging for Mac OS X Failure
+                uint errno = .errno;
+                import std.exception, std.stdio;
+                writeln("--------FAILURE----------");
+                writeln(tempDir);
+                writeln(filename);
+                writeln(errnoString(errno));
+                writeln("-------------------------");
+
+                continue;
+            }
+            else
+            {
+                import std.stdio;
+                writeln("--------SUCCESS----------");
+                writeln(tempDir);
+                writeln(filename);
+                writeln("-------------------------");
+            }
+        }
+        else version(Windows)
+        {
+            auto fd = CreateFileW(tempCString!FSChar(filename), GENERIC_WRITE, 0, null, CREATE_NEW,
+                                  FILE_ATTRIBUTE_NORMAL | FILE_FLAG_SEQUENTIAL_SCAN, HANDLE.init);
+            if (fd == INVALID_HANDLE_VALUE)
+                continue;
```

Comment:
The loop should be unconditionally retried on EEXIST or equivalent. Other errors, like ENOPERM or ENOENT (temporary directory inaccessible on nonexisting) should throw immediately.

Reply:
Fixed.